File: sdk/py/splendor_actor/types/color.py
from enum import Enum
from typing import List, Union


class Color(Enum):
    """
    Enum for the colors of the tokens and cards
    """

    BLACK = 0
    BLUE = 1
    GREEN = 2
    RED = 3
    WHITE = 4
    YELLOW = 5

    @staticmethod
    def from_str(s: str) -> "Color":
        """
        Convert a string to a Color enum
        """
        s = s.lower()
        if s == "black":
            return Color.BLACK
        elif s == "blue":
            return Color.BLUE
        elif s == "green":
            return Color.GREEN
        elif s == "red":
            return Color.RED
        elif s == "white":
            return Color.WHITE
        elif s == "yellow":
            return Color.YELLOW
        else:
            raise ValueError(f"Invalid color string: {s}")

    def __str__(self):
        return self.name

    def __repr__(self):
        return self.name

    def to_json(self) -> str:
        """
        Convert a Color to a json
        """
        return self.name.lower()


class ColorVec:
    """
    represent the color combinations
    """

    vec: List[int]

    def __init__(self, vec: List[int]):
        assert len(vec) == 6, "ColorVec should have 6 elements"

        self.vec = vec

    def __getitem__(self, index: Union[int, Color]) -> int:
        if isinstance(index, int):
            return self.vec[index]
        else:
            return self.vec[index.value]

    def __setitem__(self, index: Union[int, Color], value: int):
        assert value >= 0, "ColorVec should not contain negative values"

        if isinstance(index, int):
            self.vec[index] = value
        else:
            self.vec[index.value] = value

    def __lt__(self, other):
        return self.vec < other.vec

    def __le__(self, other):
        return self.vec <= other.vec

    def __gt__(self, other):
        return self.vec > other.vec

    def __ge__(self, other):
        return self.vec >= other.vec

    def __eq__(self, other):
        return self.vec == other.vec

    def __ne__(self, other):
        return self.vec != other.vec

    def __iter__(self):
        return ((Color(color), count) for color, count in enumerate(self.vec))

    def __add__(self, other):
        return ColorVec([a + b for a, b in zip(self.vec, other.vec)])

    def __sub__(self, other):
        return ColorVec([a - b for a, b in zip(self.vec, other.vec)])

    def saturating_sub(self, other):
        """
        Subtract other from self, but never go below 0
        """
        return ColorVec([max(a - b, 0) for a, b in zip(self.vec, other.vec)])

    def total(self):
        """
        Return the total number of tokens in the vector
        """
        return sum(self.vec)

    @classmethod
    def empty(cls) -> "ColorVec":
        """
        Return an empty color vector
        """
        return cls([0, 0, 0, 0, 0, 0])

    def to_json(self) -> List[int]:
        """
        Convert a ColorVec to a json
        """
        return self.vec

File: sdk/go/types/color.go
package types

import (
	"encoding/json"
)

type Color uint8

const (
	BLACK  Color = 0
	BLUE   Color = 1
	GREEN  Color = 2
	RED    Color = 3
	WHITE  Color = 4
	YELLOW Color = 5
)

var colorNames = map[Color]string{
	BLACK:  "black",
	BLUE:   "blue",
	GREEN:  "green",
	RED:    "red",
	WHITE:  "white",
	YELLOW: "yellow",
}

var colorValues = map[string]Color{
	"black":  BLACK,
	"blue":   BLUE,
	"green":  GREEN,
	"red":    RED,
	"white":  WHITE,
	"yellow": YELLOW,
}

func (c Color) MarshalJSON() ([]byte, error) {
	if name, ok := colorNames[c]; ok {
		return json.Marshal(name)
	}
	return nil, ErrInvalidColor
}

func (c *Color) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}

	if value, ok := colorValues[s]; ok {
		*c = value
		return nil
	}
	return ErrInvalidColor
}

// ColorVec represent the color combinations
type ColorVec [6]uint8

func NewColorVec() *ColorVec {
	return &ColorVec{}
}

func (cv *ColorVec) Add(other *ColorVec) {
	for i := 0; i < 6; i++ {
		cv[i] += other[i]
	}
}

func (cv *ColorVec) Sub(other *ColorVec) {
	for i := 0; i < 6; i++ {
		cv[i] -= other[i]
	}
}

func (cv *ColorVec) SaturatingSub(other *ColorVec) {
	for i := 0; i < 6; i++ {
		if cv[i] < other[i] {
			cv[i] = 0
		} else {
			cv[i] -= other[i]
		}
	}
}

func (cv *ColorVec) Total() uint8 {
	var total uint8
	for i := 0; i < 6; i++ {
		total += cv[i]
	}
	return total
}
